javascript_nodejs/10.prompt-validations/dialogs/mainDialog/index.js

```diff
@@ -0,0 +1,83 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+const { DialogSet } = require('botbuilder-dialogs');
+
+const OnboardingDialog = require('../onboard');
+
+const DIALOG_STATE_PROPERTY = 'dialogState';
+
+const USER_PROFILE_PROPERTY = 'user';
+
+const ONBOARD_USER = 'onboard_user';
+
+class MainDialog {
+    /**
+     * 
+     * @param {ConversationState} conversationState A ConversationState object used to store dialog state.
+     * @param {UserState} userState A UserState object used to store user profile information.
+     */
+    constructor (conversationState, userState) {
+
+        // Creates a new state accessor property. 
+        // See https://aka.ms/about-bot-state-accessors to learn more about bot state and state accessors.
+        this.conversationState = conversationState;
+        this.userState = userState;
+        
+        // Create a property used to store dialog state.
+        this.dialogState = this.conversationState.createProperty(DIALOG_STATE_PROPERTY);
+
+        // Create a dialog set to include the dialogs used by this bot.
+        this.dialogs = new DialogSet(this.dialogState);
+
+        // Create some properties used to store values from the user.
+        this.userProfile = this.userState.createProperty(USER_PROFILE_PROPERTY)
+
+        // Create the main user onboarding dialog.
+        this.dialogs.add(new OnboardingDialog(ONBOARD_USER, this.userProfile));
+    }
+
+
+    /**
+     * 
+     * @param {TurnContext} context A TurnContext object representing an incoming message to be handled by the bot.
+     */
+    async onTurn(context) {
+        // See https://aka.ms/about-bot-activity-message to learn more about the message and other activity types.
+        if (context.activity.type === 'message') {
+
+            // Create dialog context
+            const dc = await this.dialogs.createContext(context);
+
+            const utterance = (context.activity.text || '').trim().toLowerCase();
+            if (utterance === 'cancel') { 
+                if (dc.activeDialog) {
+                    await dc.cancelAll();
+                    await dc.context.sendActivity(`Ok... Cancelled.`);
+                } else {
+                    await dc.context.sendActivity(`Nothing to cancel.`);
+                }
+            }
+            
+            // Continue the current dialog if one is pending.
+            if (!context.responded) {
```

Comment:
This if-block and comment aren't necessary. `context.responded` is `true` only when the bot responds to a cancel request. When the bot handles a cancel request, the dialog stack is always empty and `dc.continue()` doesn't do anything noteworthy when it's called with an empty dialog stack.